File: proton-c/src/types/array.c
/*
 *
 * Licensed to the Apache Software Foundation (ASF) under one
 * or more contributor license agreements.  See the NOTICE file
 * distributed with this work for additional information
 * regarding copyright ownership.  The ASF licenses this file
 * to you under the Apache License, Version 2.0 (the
 * "License"); you may not use this file except in compliance
 * with the License.  You may obtain a copy of the License at
 *
 *   http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing,
 * software distributed under the License is distributed on an
 * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
 * KIND, either express or implied.  See the License for the
 * specific language governing permissions and limitations
 * under the License.
 *
 */

#include <proton/util.h>
#include <arpa/inet.h>
#include <stdio.h>
#include <stdlib.h>
#include "encodings.h"
#include "value-internal.h"

static char type_to_code(enum TYPE type)
{
  switch (type)
  {
  case EMPTY: return 'n';
  case BYTE: return 'b';
  case UBYTE: return 'B';
  case SHORT: return 'h';
  case USHORT: return 'H';
  case INT: return 'i';
  case UINT: return 'I';
  case LONG: return 'l';
  case ULONG: return 'L';
  case FLOAT: return 'f';
  case DOUBLE: return 'd';
  case CHAR: return 'C';
  case SYMBOL: return 's';
  case STRING: return 'S';
  case BINARY: return 'z';
  case LIST: return 't';
  case MAP: return 'm';
  default: return -1;
  }
}

static uint8_t type_to_amqp_code(enum TYPE type)
{
  switch (type)
  {
  case EMPTY: return PNE_NULL;
  case BOOLEAN: return PNE_BOOLEAN;
  case BYTE: return PNE_BYTE;
  case UBYTE: return PNE_UBYTE;
  case SHORT: return PNE_SHORT;
  case USHORT: return PNE_USHORT;
  case INT: return PNE_INT;
  case UINT: return PNE_UINT;
  case CHAR: return PNE_UTF32;
  case LONG: return PNE_LONG;
  case ULONG: return PNE_ULONG;
  case FLOAT: return PNE_FLOAT;
  case DOUBLE: return PNE_DOUBLE;
  case SYMBOL: return PNE_SYM32;
  case STRING: return PNE_STR32_UTF8;
  case BINARY: return PNE_VBIN32;
  case LIST: return PNE_LIST32;
  case MAP: return PNE_MAP32;
  case ARRAY: return PNE_ARRAY32;
  default:
    pn_fatal("no amqp code for type: %i", type);
    return -1;
  }
}

pn_array_t *pn_array(enum TYPE type, int capacity)
{
  pn_array_t *l = malloc(sizeof(pn_array_t) + capacity*sizeof(pn_value_t));
  if (l) {
    l->type = type;
    l->capacity = capacity;
    l->size = 0;
  }
  return l;
}

int pn_array_add(pn_array_t *a, pn_value_t v)
{
  if (a->capacity <= a->size) {
    fprintf(stderr, "wah!\n");
    return -1;
  }

  a->values[a->size++] = v;
  return 0;
}

void pn_free_array(pn_array_t *a)
{
  free(a);
}

void pn_visit_array(pn_array_t *a, void (*visitor)(pn_value_t))
{
  for (int i = 0; i < a->size; i++)
  {
    pn_visit(a->values[i], visitor);
  }
}

size_t pn_format_sizeof_array(pn_array_t *array)
{
  size_t result = 4;
  for (int i = 0; i < array->size; i++)
  {
    result += pn_format_sizeof(array->values[i]) + 2;
  }
  return result;
}

int pn_format_array(char **pos, char *limit, pn_array_t *array)
{
  int e;
  if ((e = pn_fmt(pos, limit, "@%c[", type_to_code(array->type)))) return e;
  if ((e = pn_format_value(pos, limit, array->values, array->size))) return e;
  if ((e = pn_fmt(pos, limit, "]"))) return e;
  return 0;
}

size_t pn_encode_sizeof_array(pn_array_t *array)
{
  size_t result = 9;
  for (int i = 0; i < array->size; i++)
  {
    // XXX: this is wrong, need to compensate for code
    result += pn_encode_sizeof(array->values[i]);
  }
  return result;
}

size_t pn_encode_array(pn_array_t *array, char *out)
{
  // code
  out[0] = (uint8_t) PNE_ARRAY32;
  // size will be backfilled
  // count
  *((uint32_t *) (out + 5)) = htonl(array->size);
  // element code
  out[9] = (uint8_t) type_to_amqp_code(array->type);

  char *vout = out + 10;
  for (int i = 0; i < array->size; i++)
  {
    char *codeptr = vout - 1;
    char codeval = *codeptr;
    vout += pn_encode(array->values[i], vout-1) - 1;
    *codeptr = codeval;
  }

  // backfill size
  *((uint32_t *) (out + 1)) = htonl(vout - out - 5);
  return vout - out;
}

File: proton-c/bindings/ruby/lib/types/array.rb
#--
# Licensed to the Apache Software Foundation (ASF) under one
# or more contributor license agreements.  See the NOTICE file
# distributed with this work for additional information
# regarding copyright ownership.  The ASF licenses this file
# to you under the Apache License, Version 2.0 (the
# "License"); you may not use this file except in compliance
# with the License.  You may obtain a copy of the License at
#
#   http://www.apache.org/licenses/LICENSE-2.0
#
# Unless required by applicable law or agreed to in writing,
# software distributed under the License is distributed on an
# "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
# KIND, either express or implied.  See the License for the
# specific language governing permissions and limitations
# under the License.
#++

#--
# Patch the Array class to provide methods for adding its contents
# to a Qpid::Proton::Data instance.
#++

module Qpid::Proton::Types

  # Holds the information for an AMQP Array compound type.
  #
  # It holds the type for the array and the descriptor if the
  # array is described.
  #
  # @private
  #
  class ArrayHeader
    attr_reader :type
    attr_reader :descriptor

    def initialize(type, descriptor = nil)
      @type = type
      @descriptor = descriptor
    end

    # Returns true if the array is described.
    def described?
      !@descriptor.nil?
    end

    def ==(that)
      ((@type == that.type) && (@descriptor == that.descriptor))
    end
  end

end

# @private
class Array # :nodoc:

  # Used to declare an array as an AMQP array.
  #
  # The value, if defined, is an instance of Qpid::Proton::Types::ArrayHeader
  attr_accessor :proton_array_header

  # Returns true if the array is the a Proton described type.
  def proton_described?
    !@proton_array_header.nil? && @proton_array_header.described?
  end

  # Puts the elements of the array into the specified Qpid::Proton::Data object.
  def proton_put(data)
    raise TypeError, "data object cannot be nil" if data.nil?

    if @proton_array_header.nil?
      proton_put_list(data)
    else
      proton_put_array(data)
    end
  end

  private

  def proton_put_list(data)
    # create a list, then enter it and add each element
    data.put_list
    data.enter
    each do |element|
      # get the proton type for the element
      mapping = Qpid::Proton::Codec::Mapping.for_class(element.class)
      # add the element
      mapping.put(data, element)
    end
    # exit the list
    data.exit
  end

  def proton_put_array(data)
    data.put_array(@proton_array_header.described?, @proton_array_header.type)
    data.enter
    if @proton_array_header.described?
      data.symbol = @proton_array_header.descriptor
    end

    each do |element|
      @proton_array_header.type.put(data, element)
    end

    data.exit
  end

  class << self

    # Gets the elements of an array or list out of the specified
    # Qpid::Proton::Data object.
    def proton_get(data)
      raise TypeError, "can't convert nil into Qpid::Proton::Data" if data.nil?

      type = data.type

      if type == Qpid::Proton::Codec::LIST
        result = proton_get_list(data)
      elsif type == Qpid::Proton::Codec::ARRAY
        result = proton_get_array(data)
      else
        raise TypeError, "element is not a list and not an array"
      end
    end

    private

    def proton_get_list(data)
      size = data.list
      raise TypeError, "not a list" unless data.enter
      elements = []
      (0...size).each do
        data.next
        type = data.type
        raise TypeError, "missing next element in list" unless type
        elements << type.get(data)
      end
      data.exit
      return elements
    end

    def proton_get_array(data)
      count, described, type = data.array

      raise TypeError, "not an array" unless data.enter
      elements = []

      descriptor = nil

      if described
        data.next
        descriptor = data.symbol
      end

      elements.proton_array_header = Qpid::Proton::Types::ArrayHeader.new(type, descriptor)
      (0...count).each do |which|
        if data.next
          etype = data.type
          raise TypeError, "missing next element in array" unless etype
          raise TypeError, "invalid array element: #{etype}" unless etype == type
          elements << type.get(data)
        end
      end
      data.exit
      return elements
    end

  end

end

